core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,873 @@
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let exploration = sampleExplorationDict;
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', '', '', []), 1, 0, 'default_outcome'
+      );
+      answerClassificationResult.outcome.labelledAsCorrect = true;
+
+      spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+      spyOn(contextService, 'getExplorationId')
+        .and.returnValue('explorationId1');
+      spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+      spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+      spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+        .and.returnValue(Promise.resolve(exploration));
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+      spyOn(expressionInterpolationService, 'processHtml')
+        .and.callFake((html, envs) => html);
+      questionPlayerEngineService.init(
+        multipleQuestionsBackendDict, successHandler, failHandler);
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(questionPlayerEngineService.getCurrentIndex()).toBe(0);
+
+      questionPlayerEngineService.recordNewCardAdded();
+
+      expect(questionPlayerEngineService.getCurrentIndex()).toBe(1);
+    });
+
+    it('should not create next question if the existing ' +
+      'question is the last one', () => {
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', '', '', []), 1, 0, 'default_outcome'
+      );
+      answerClassificationResult.outcome.labelledAsCorrect = true;
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+      spyOn(expressionInterpolationService, 'processHtml')
+        .and.callFake((html, envs) => html);
+      spyOn(focusManagerService, 'generateFocusLabel')
+        .and.returnValue('focusLabel');
+
+      questionPlayerEngineService.init(
+        [singleQuestionBackendDict], successHandler, failHandler);
+
+      let nextQuestion = null;
+
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(successCallback).toHaveBeenCalledWith(
```

Comment:
What does all this have to do with the description of the test?

Reply:
In `submitAnswer()` nextCard ( ~ next question) is a local variable, every time, if the correct answer has submitted this function, will load the next question state and creates a card, and pass it into the callback function.
If the nextCard/nextQuestion is null no further question will be loaded.

There is a better/more readable way to test this.
We can `expect(stateCardObjectFactory.createNewCard()).not.toHaveBeenCalled())`
but I'm afraid that you have explicitly mentioned not use expect any function/Spies to be called as it is testing the implementation rather than interface.

Reply:
In this case it is fine to do as you suggest. Just make sure the same expectation happens earlier on with a toHaveBeenCalled() expectation, to show that there is a difference.

Note that, in this case, you are taking "createNewCard()" to be the actual condition being tested (the test description is literally "should not create new question") so this _is_ the interface.

Reply:
done!